Feat/ddr4 interface calculation

marcomoerz opened the pull request:
This pull request introduces calculations for the DDR4 interface, encompassing both preamble and postamble calculations. Additionally, it includes initial enhancements in the bus and command pattern encoder to accommodate DRAM-specific logic values such as V or X. Furthermore, comments have been incorporated into the DDR4 implementation to enhance readability.

Commit:
Added basic ddr4 interface calculation

The supported pins are: clock, DQS, DQ and command bus. The readability of the code was improved. DDR4 now supports x4, x8 and x16 devices and scales the interface calculation accordingly.

Commit:
PrePostamble implemented

- PrePostamble interface calculation added
- Test for counters
- Added custom values for command pattern generation fields V and X (L, H, last value)
- Added PrePostamble field in ddr.json
- Removed unused code

Commit:
expanded interface calculation

- PrePostamble interface calculation added for seamless PrePostamble
- bus default idle setting added
- patternencoder setting for X, V, AP and BL added
- bug fixes

Commit:
interface calculation clock energy test

- bug fixes
- expanded interface test
- moved count DQs pairs in interface calculation

Commit:
added energy tests

- DQs energy
- DQ energy
- CA energy
- Preamble and Postamble energy

- Pattern 3 and Pattern 4 added

Commit:
differential pairs in stats

- The differential pair stats (zeroes, ones, ...) of for example dqs include now all lines.
  For example:
    - for a x16 device the DQs zeroes include all 4 lanes
- bug fixes
- clearer comments

Commit:
bug fixes

- added default return value false for applyBitSpec in Pattern.h
- removed use of designated initializers for structs

Commit:
bug fixes

- removed use of designated initializers for
  structs in test_bus.cpp

Commit:
bug fixes

removed use of designated initializers for
test_ddr_base.cpp

Commit:
Refactor pattern encoder settings in DDR standards

- changed BusSettings
- PatternEncoder Settings override non target coordinate pattern_descriptors
- overrides stored in unordered_map

Commit:
add include unordered_map

Commit:
Update bus idle pattern and init pattern handling

- bus idle pattern and init pattern now support bus widths greater than 64
- added extended bus tests

Commit:
Remove memSpec.numberOfDevices in DDR4

Commit:
Add include for <array> in test_bus_extended.cpp

lsteiner-tukl commented on src/DRAMPower/DRAMPower/standards/ddr4/core_calculation_DDR4.h:
@IESE-T3 Bringt manuelles inlinen überhaupt was?

marcomoerz replied:
Ja manuelles Inlineing kann die Leistung verbessern und fordert den Compiler auf die Funktion zu inlinen. Somit ist gegeben, dass die Funktion im kompilierten Binary inlined wurde. Ohne inline ist dies abhängig von der Optimierung.

lsteiner-tukl commented on src/DRAMPower/DRAMPower/util/bus.h:
```diff
@@ -31,70 +34,170 @@ struct bus_stats_t {
 class Bus {
+
+private:
+	enum class __BusInitPatternSpec
```
Ich glaube __ sollte nur in der Standardbibliothek genutzt werden.

marcomoerz replied:
Abgeändert zu BusInitPatternSpec_

Commit:
Updated pattern encoder

- added pattern encoder init
- test for pattern encoder

Commit:
added clang format and clang tidy files

Commit:
Reworked bus stats

use virtual function in dram_base for init encoder pattern
added bus tests for extended bus

Commit:
Update commandBus.zeroes values in test_interface files

Improved interface calculation to align with the new bus implementation.

Commit:
Update buswidth and bus_array_size in test_bus_extended.cpp

Commit:
fix gcc build error

Commit:
Update buswidth and bus_array_size constants

Commit:
Add static power function for prepostamble calculation
